### ISSUE TITLE
Correct way to import apollo-server (rather than require)

### ISSUE THREAD

**Issue by meteorpublish** [apollographql/apollo-server#1356]
I'm trying to import 'apollo-server' using `--experimental-modules` in Node 10. What is the correct syntax?

```js
import { ApolloServer } from 'apollo-server';
// SyntaxError: The requested module 'apollo-server' does not provide an export named 'ApolloServer'
import ApolloServer from 'apollo-server'
// IDE warning: "Default export is not declared in imported module
```

**Comment by evans** [comment]
@meteorpublish Super odd, https://github.com/apollographql/apollo-server/blob/95f4d5e00b0d567a34463d7b27a9fdc3dfdf6eea/packages/apollo-server/src/index.test.ts#L7 uses the first style of import. Does Node 10 require that the package.json place the entry point somewhere other than main?

**Comment by TimPerry** [comment]
+1 on this, also trying to use the flag with node 10.

**Comment by dandv** [comment]
Possibly related: https://github.com/apollographql/graphql-tools/issues/913#issuecomment-409694650

**Comment by andreisocaciu** [comment]
Any news about support on node 12 with --experimental-modules flag on?

**Comment by trevorblades** [comment]
The correct syntax is this one:

```js
import {ApolloServer} from 'apollo-server'; // or apollo-server-[your framework]
```

I was able to get this to work using https://github.com/standard-things/esm. Closing this issue now, please reopen if you have more questions!

**Comment by rolfen** [comment]
> The correct syntax is this one:
> 
> ```js
> import {ApolloServer} from 'apollo-server'; // or apollo-server-[your framework]
> ```
> 
> I was able to get this to work using https://github.com/standard-things/esm. Closing this issue now, please reopen if you have more questions!

I was not - it complains that 
```
SyntaxError: The requested module 'file:///home/me/code/test/node_modules/apollo-server/dist/index.js' does not provide an export named 'ApolloServer'
```

Just saying.

**Comment by trevorblades** [comment]
@rolfen what version of `apollo-server` are you using?

**Comment by rolfen** [comment]
> @rolfen what version of `apollo-server` are you using?

apollo-server@2.8.1
node v12.7.0

**Comment by trevorblades** [comment]
That's interesting that changing the extension would fix it. Normally, using `esm` allows you to use `import`/`export` syntax in .js files. Here's a codesandbox I put together showing an example ESM + Apollo Server setup: https://codesandbox.io/s/polished-http-09vyl

**Comment by djfarrelly** [comment]
Using `"type": "module"` in `package.json` with Node 12 and `--experimental-modules` I had to use this workaround:
```js
import apollo from 'apollo-server'
const { ApolloServer } = apollo
```
Instead of what I assumed would work:
```js
import { ApolloServer } from 'apollo-server'
```
Which throws the `The requested module 'apollo-server' does not provide an export named 'ApolloServer'` as others mentioned above. I wanted to share this workaround in case anyone else is running into the same issues using the `--experimental-modules` flag or using modules w/ node 13 w/out the flag ([docs](https://nodejs.org/api/esm.html)).

**Comment by dandv** [comment]
Would be nice to have proper ES Modules support (without `esm`), now that they're no longer experimental since Node v13.2.0.

@trevorblades should this issue be reopened?

**Comment by trevorblades** [comment]
@dandv @djfarrelly it looks like we could support ES modules and CommonJS at the same time by following a pattern like [the ones mentioned in the NodeJS API reference](https://nodejs.org/api/esm.html#esm_writing_dual_packages_while_avoiding_or_minimizing_hazards).

That being said, Apollo Server is written in TypeScript and compiled to be compatible with CommonJS. We're [passing the `esModuleInterop` flag to the TS compiler](https://github.com/apollographql/apollo-server/blob/master/tsconfig.base.json#L7), so it's surprising that named exports aren't working as expected using native NodeJS ES modules. I found a few issues on the TypeScript repo that might relate to this:

- https://github.com/microsoft/TypeScript/issues/21621
- https://github.com/microsoft/TypeScript/issues/27293

If there was a way to replicate the pattern identified in the NodeJS reference above using TypeScript config options, that would be ideal.

**Comment by platocrat** [comment]
Still getting an error when using Node v14.4.0

https://github.com/apollographql/apollo-server/issues/3786#issuecomment-652694891

**Comment by boblitex** [comment]
`The requested module 'apollo-server-express' is expected to be of type CommonJS, which does not support named exports. CommonJS modules can be imported by importing the default export`.

I tried to import ApolloServer using Es6 syntax and the above was the error output. something isn't right with that still. 
using Node 14.8.0 and module support is enabled

**Comment by thobiasn** [comment]
Getting this error aswell. No luck installing esm, renaming to .mjs, adding type: module to my package.json or installing the @rc package.

**Comment by dandv** [comment]
I've managed to solve this exact problem (named ESM imports from the TypeScript output), including the error message reported by @boblitex, by using [conditional exports](https://nodejs.org/api/esm.html#esm_conditional_exports), which [launched in Node 13.7.0, and have been no longer experimental since v13.10](https://alexomara.com/blog/dual-packages-in-node-js-conditional-exports/).

You can see a [diff](https://github.com/dandv/local-iso-dt/commit/00d02af5fcaf06ca17c3f2b9e86f44ddc174b628) implementing this for a package I'm maintaining.

What needs to be done, essentially, is to output `.mjs` files (might require manual renames [for now](https://github.com/microsoft/TypeScript/issues/18442)) and add the following to the various `package.json` files:

```json
"exports": {
  "node": {
    "import": "./index.mjs",
    "require": "./index.js"
  },
"default": "./index.mjs"
},
```

@trevorblades

**Comment by jaibatrik** [comment]
Shouldn't we keep this issue open for tracking since the problem still persists?

**Comment by mushketyk** [comment]
@trevorblades This issue still persists with the `3.0.2` version of `apollo-server`. Should it be reopened?

**Comment by trevorblades** [comment]
@mushketyk I'm doing `import {ApolloServer} from 'apollo-server'` with Apollo Server 3, Node 14, and `"type": "module"` in my package.json. Works perfectly fine as far as I can tell. Can you describe the issue you're seeing?

**Comment by glasser** [comment]
Let's keep this issue closed because it's a duplicate of #3786. @trevorblades cool to hear it's working for you, though that's surprising. Should we close #3786?
